### PR TITLE
distributor: bound rebatching preallocation

### PR DESCRIPTION
**Pull request by carles-grafana**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: modules/distributor/distributor.go [hunk 1]
```diff
@@ -48,6 +48,9 @@ const (
 	distributorRingKey = "distributor"
 
 	truncationLogsPerSecond = 1
+
+	// 16 covers the typical spans-per-trace seen in production.
+	maxPreallocSpansPerTrace = 16
```
Comment (oleg-kozlyuk-grafana) [inline modules/distributor/distributor.go:53]: Allocations are expensive per-unit, not so much per-byte - it may be better to size to p90 or p95 than to p50.

File: modules/distributor/distributor.go [hunk 2]
```diff
@@ -662,10 +665,20 @@ func (d *Distributor) sendToKafka(ctx context.Context, userID string, keys []uin
...
 	truncatedCount := truncatedAttributesCount{}
 
+	// Estimate spans per trace, but cap it: an uncapped spanCount/tracesPerBatch
+	// would preallocate too much memory for high-cardinality requests.
+	perTracePrealloc := spanCount / tracesPerBatch
+	if perTracePrealloc > maxPreallocSpansPerTrace {
+		perTracePrealloc = maxPreallocSpansPerTrace
+	}
+	if perTracePrealloc < 1 {
+		perTracePrealloc = 1
```
Comment (oleg-kozlyuk-grafana) [inline modules/distributor/distributor.go:679]: Perhaps still 2-4?

**Review by oleg-kozlyuk-grafana — APPROVED** [review]
Approved with nitpicks - might benefit from benching/tuning (unless already done?)

**Comment by oleg-kozlyuk-grafana** [comment]
BTW, any regressions in benchmarks?

**Comment by carles-grafana** [comment]
I ran the rebatching benchmarks against base and this PR.

High-cardinality improves substantially:
- 5k traces x 1 span: ~22MB/op -> ~2.8MB/op, ~4.6ms -> ~1.85ms
- 20k traces x 1 span: ~334MB/op -> ~11MB/op, ~39ms -> ~7.9ms

Normal prod-like traffic is effectively unchanged

I also tested raising the minimum prealloc to 4. It only affects tiny requests (<~80 spans). It reduces allocs for small multi-span traces, but slightly over-allocates 1-span requests and does not affect normal/large/high-cardinality cases.

File: modules/distributor/distributor.go (hunk 1 above)
Comment (carles-grafana) [reply modules/distributor/distributor.go:53]: 16 is not a measured p50/p95, it’s the smallest conservative cap I found that keeps normal benchmark allocation counts unchanged while still bounding the high-cardinality case.
